enclave_build/src/cache.rs

```diff
@@ -0,0 +1,219 @@
+/// Name of the cache index file which stores the (image URI <-> image hash) mappings.
+pub const CACHE_INDEX_FILE_NAME: &str = "index.json";
+/// The name of the OCI layout file from the cache.
+pub const CACHE_OCI_LAYOUT_FILE: &str = "oci-layout";
+
+/// Constants used for complying with the OCI cache structure
+pub const DEFAULT_MEDIA_TYPE: &str = "application/vnd.oci.image.manifest.v1+json";
+pub const REF_ANNOTATION: &str = "org.opencontainers.image.ref.name";
+pub const OCI_LAYOUT: (&str, &str) = ("imageLayoutVersion", "1.0.0");
+
+/// Manifest entry in index file
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ManifestEntry {
+    #[serde(rename = "mediaType")]
+    media_type: String,
+    #[serde(rename = "size")]
+    size: usize,
+    #[serde(rename = "digest")]
+    digest: String,
+    #[serde(rename = "annotations")]
+    annotations: HashMap<String, String>,
+}
+
+/// Cache index file structure
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CacheIndex {
+    #[serde(rename = "schemaVersion", default = "schema_default")]
+    schema_version: u32,
+    #[serde(rename = "manifests", default = "Vec::new")]
+    manifests: Vec<ManifestEntry>,
+}
+
+impl Default for CacheIndex {
+    fn default() -> Self {
+        CacheIndex {
+            schema_version: 2,
+            manifests: Vec::new(),
+        }
+    }
+}
+
+fn schema_default() -> u32 {
+    2
+}
+
+/// Struct which provides operations with the local cache.
+///
+/// The index file is located in the cache root folder and keeps track of images stored in cache.
+///
+/// The cache structure is:
+///
+/// {CACHE_ROOT_PATH}/index.json\
+/// {CACHE_ROOT_PATH}/blobs/sha256\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_1\
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_2\
+/// etc.
+///
+/// A blob can be:
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct CacheManager {
+    /// The root folder of the cache
+    root_path: PathBuf,
+
+    /// A map storing the cached images, with the map entry format being (image_reference, image_hash)
+    cached_images: HashMap<String, String>,
+}
+
+impl CacheManager {
+    /// Creates a new CacheManager instance and returns it. As argument, a path to the root folder
+    /// of the cache should be provided.
+    ///
+    /// Apart from that, the function also creates (if not already created) all folders from the path
+    /// specified as argument.
+    ///
+    /// If an index file exists at the path, it loads the file's contents into the 'cached_images'
+    /// field. If not, a new empty index file is created at that path.
+    pub fn new(root_path: &Path) -> Result<Self> {
```

Comment:
Does it help to use `Path` instead of `PathBuf`? We would avoid all the `to_path_buf`s.